### DISCUSSION TITLE
Reading Cycle time of DB Read

### DISCUSSION THREAD

**Discussion by dattatrayshindebuck** [fbarresi/Sharp7#25] [Q&A] [answered]
Hello,

I am developing an HMI for my PLC using Sharp7 in C# based WPF Application.

I want to read values from my DB in Siemens PLC in each 50 milliseconds if possible. I have found that my DB read command is taking time between 20ms to 100ms. Also, this time is different on each cycle. 

I would like to know why is this happening ? 

The same function trying to read same values from PLC is taking variable time. Here is the plotted graph of time taken to read vs 30 seconds time frame. 

![ReadCycleTimeSharp7](https://user-images.githubusercontent.com/82153465/125782288-7fb62bfb-d433-4ea6-820f-479c43708a2d.JPG)

**Comment by fbarresi** [comment, marked as answer]
Hi!

Thank you for starting a discussion here!

In my opinion, reaching 50 ms of update rate may be difficult. 
Sharp7 is build on the put/get functionality of every siemens PLC and this protocol is not real time. 
A read duration up to 100 ms doesn't surprise me: there are many good reasons why this process may last longer than initially expected for example:
- PLC cycle time
- network traffic 
- amount of data
- other readings over the same socket

I think you have two way to go further: *live with it* or *try to optimize*.

### Live with it
Your application can't practically reach a update rate better than 100 ms. If you can accept it, it's all right!
I don't know if you measured just one variable or more than one, but since Sharp7 is not thread-safe, you will experience, that reading many variables will additionally slow down the whole application. So for a complex HMI I can image you will probably reach 200 or 300 ms update rate. It's not great, but even betten than WinCC in many use cases!

### Try to optimize
I see 2 main ways for speeding up such a process.
1. Be clever than a PLC: As I told you, the amount of data effect the transmission speed. So instead of polling data every 50 ms, you may let the PLC itself doing this job for you and generate a flag (BOOL) every time data changed.
In this case you just read the data once at the beginning and every time you see the flag (over a polling task). Otherwise, if your data really changes every 50 ms, you would not reach anything with this approach. On the other hand let's assume you have many different data you want to observe, you may think that this approach does not scale, but it does! The protocol also offers you a way to create batches (multivars) for all the flags you would like to observe, so in this case they would run synchronously.

2.  Open a dedicated channel: You can start an additional connection with the same PLC (just a new sharp7 client) dedicated to stuff that have to reach an high update rate. Older PLC models are limited in the number of simultanes connections, but the new ones not. Of course you may combine the approach 1. also here and achieve better results. I would advise you also to check if the reading requests are not delayed by incorrect calls: if a reading is still not finished should no new one starts.

I hope I could help you. 
Feel free to continue the discussion or share your achievements here. I would be really happy to know if you have accomplished the goal!

Best regards,
FB
